src/Compilers/CSharp/Portable/CSharpResources.resx

```diff
@@ -5247,4 +5246,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_DefaultInPattern" xml:space="preserve">
     <value>A default literal 'default' is not valid as a pattern. Use another literal (e.g. '0' or 'null') as appropriate. To match everything, use a discard pattern 'var _'.</value>
   </data>
-</root>
+  <data name="IDS_FeatureEnumGenericTypeConstraint" xml:space="preserve">
```

Comment:
Consider adding a single feature for all 3 constraints.
Then it may make sense to report a custom error message for this feature, when using an older LangVersion (as it was done for ERR_NamedArgumentSpecificationBeforeFixedArgument ).
For example, "System.Enum, System.Delegate and unmanaged constraints are not available in C# {0} . Please use language version {1} or greater to use such constraints."

The only downside is that this new error code will need to be added the list monitored by UpgradeProject.

Reply:
I thought about it, and it looks to me like although these features are logically related, set of types using them don't normally intersect. I'd prefer to keep the messages separate, so that you just get the feature you're typing right now in the diagnostic.

---
In reply to: [162500344](https://github.com/dotnet/roslyn/pull/24199#discussion_r162500344) [](ancestors = 162500344)